src/debugAdapter/goDebug.ts

```diff
@@ -479,71 +479,69 @@ class Delve {
 			});
 		});
 	}
 
-	close(): Thenable<void> {
+	/**
+	 * To close the debugging session we do the following; If its a local process,
+	 * we try to halt and detach within a timeout otherwise we kill the debugger. If its a remote
+	 * process we try to halt and restart the remote target process running under the debugger.
+	 */
+	public close(): Thenable<void> {
 		if (this.noDebug) {
 			// delve isn't running so no need to halt
 			return Promise.resolve();
 		}
 		log('HaltRequest');
 
-		return new Promise(resolve => {
-			let timeoutToken: NodeJS.Timer;
-			if (this.debugProcess) {
-				timeoutToken = setTimeout(() => {
-					log('Killing debug process manually as we could not halt and detach delve in time');
-					killTree(this.debugProcess.pid);
-					resolve();
-				}, 1000);
-			}
+		return new Promise(async resolve => {
+			const timeoutToken: NodeJS.Timer = !this.isRemoteDebugging() && setTimeout(() => {
+				log('Killing debug process manually as we could not halt delve in time');
+				killTree(this.debugProcess.pid);
+				resolve();
+			}, 1000);
 
-			this.callPromise('Command', [{ name: 'halt' }]).then(() => {
-				if (timeoutToken) {
-					clearTimeout(timeoutToken);
-				}
+			let errMsg;
+			try {
+				await this.callPromise('Command', [{ name: 'halt' }]);
+			} catch (err) {
 				log('HaltResponse');
-				if (!this.debugProcess) {
-					log('RestartRequest');
-					return this.callPromise('Restart', this.isApiV1 ? [] : [{ position: '', resetArgs: false, newArgs: [] }])
-						.then(null, err => {
-							log('RestartResponse');
-							logError(`Failed to restart - ${(err || '').toString()}`);
-						})
-						.then(() => resolve());
-				} else {
-					log('DetachRequest');
-					return this.callPromise('Detach', [this.isApiV1 ? true : { Kill: true }])
-						.then(null, err => {
-							log('DetachResponse');
-							logError(`Killing debug process manually as we failed to detach - ${(err || '').toString()}`);
-							killTree(this.debugProcess.pid);
-						})
-						.then(() => resolve());
-				}
-			}, err => {
-				const errMsg = err ? err.toString() : '';
-				log('Failed to halt - ' + errMsg.toString());
-				if (errMsg.endsWith('has exited with status 0')) {
-					if (timeoutToken) {
-						clearTimeout(timeoutToken);
-					}
-					return resolve();
+				errMsg = err ? err.toString() : '';
+				log(`Failed to halt - ${errMsg}`);
+			}
+			clearTimeout(timeoutToken);
+
+			if (this.shouldDetach(errMsg)) {
+				log('DetachRequest');
+				try {
+					await this.callPromise('Detach', [this.isApiV1 ? true : { Kill: true }]);
+				} catch (err) {
+					log('DetachResponse');
+					logError(`The timeout will kill the debug process manually as we failed to detach - ${(err.toString() || '')}`);
 				}
-			});
+			}
+			return resolve();
 		});
 	}
+
+	private isRemoteDebugging() {
+		return !this.debugProcess;
+	}
+	private targetHasExited(errMsg: string) {
+		return errMsg.endsWith('has exited with status 0');
```

Comment:
This assumes the process has a zero exit code, I would either omit the zero or change it to a regex (if you do be sure to handle negative error codes as well).

Reply:
ehh to be fair i dont like at all checking the exited code. let me look a little if there is a better way to know this. I presume if it requires an extra call to the debug server is not the end of the world ?

Reply:
For the scope of this PR, lets keep this check as is. Its the same what we do today